### PR TITLE
Relax the assumption that the root module is FSDP wrapped

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes https://github.com/Lightning-AI/lightning/issues/18005

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/fabric/strategies/fsdp.py [hunk 1]
```diff
@@ -364,6 +371,12 @@ def save_checkpoint(
                 "`FSDPStrategy.save_checkpoint(..., storage_options=...)` is not supported because"
                 " `FSDPStrategy` does not use the `CheckpointIO`."
             )
+        if filter is not None and self._state_dict_type == "sharded":
+            # A KeyError is raised when `torch.distributed.checkpoint.load_state_dict` is called
+            raise NotImplementedError(
+                "FSDP doesn't support loading sharded filtered checkpoints, so saving them is disabled."
+            )
```
Comment (awaelchli) [inline src/lightning/fabric/strategies/fsdp.py:378]: Could you open an issue about this?

File: src/lightning/fabric/strategies/fsdp.py [hunk 2]
```diff
@@ -372,20 +385,19 @@ def save_checkpoint(
         from torch.distributed.checkpoint import FileSystemWriter, save_state_dict
         from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
 
-        modules = [module for module in state.values() if isinstance(module, FSDP)]
+        modules = [module for module in state.values() if isinstance(module, Module)]
         if len(modules) == 0:
             raise ValueError(
-                "Could not find a FSDP model in the provided checkpoint state. Please provide the model as"
+                "Could not find a model in the provided checkpoint state. Please provide the model as"
```
Comment (awaelchli) [inline src/lightning/fabric/strategies/fsdp.py:391]: could we keep the mention of FSDP models specficially?

File: src/lightning/fabric/strategies/fsdp.py [hunk 3]
```diff
@@ -372,20 +385,19 @@ def save_checkpoint(
         from torch.distributed.checkpoint import FileSystemWriter, save_state_dict
         from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
 
-        modules = [module for module in state.values() if isinstance(module, FSDP)]
+        modules = [module for module in state.values() if isinstance(module, Module)]
```
Comment (awaelchli) [inline src/lightning/fabric/strategies/fsdp.py:388]: Let's keep this check to find the modules that are FSDP wrapped. Our logic can currently only support one FSDP model + several optimizers in one state.

```suggestion
        modules = [module for module in state.values() if _has_fsdp_modules(module)]
```
and define `_has_fsdp_modules` as `any(isinstance(mod, FullyShardedDataParallel) for mod in module.modules())`

File: src/lightning/fabric/strategies/fsdp.py [hunk 4]
```diff
@@ -465,17 +477,17 @@ def load_checkpoint(
         from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
         from torch.distributed.fsdp import OptimStateKeyType
 
-        modules = {key: module for key, module in state.items() if isinstance(module, FSDP)}
+        modules = {key: module for key, module in state.items() if isinstance(module, Module)}
```
Comment (awaelchli) [inline src/lightning/fabric/strategies/fsdp.py:480]: Same here

File: src/lightning/fabric/strategies/fsdp.py [hunk 5]
```diff
@@ -218,29 +218,31 @@ def setup_module_and_optimizers(
...
-            device_id=self.root_device.index,
-            **self._fsdp_kwargs,
-        )
+        from torch.distributed.fsdp import FullyShardedDataParallel
+
+        if any(isinstance(mod, FullyShardedDataParallel) for mod in module.modules()):
+            if "auto_wrap_policy" in self._fsdp_kwargs:
+                rank_zero_warn(
+                    "A FSDP `auto_wrap_policy` is set, but the model is already wrapped. The policy will be ignored."
+                )
```
Comment (awaelchli) [inline src/lightning/fabric/strategies/fsdp.py:230]: Could you cover the warning in a test?

File: src/lightning/fabric/strategies/fsdp.py [hunk 6]
```diff
@@ -317,24 +319,29 @@ def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
...
         norm_type: Union[float, int] = 2.0,
         error_if_nonfinite: bool = True,
     ) -> Tensor:
         """Clip gradients by norm."""
-        rank_zero_warn("Gradient Clipping by Norm is currently experimental for FSDP. Proceed with Caution!")
+        from torch.distributed.fsdp.fully_sharded_data_parallel import FullyShardedDataParallel
+
+        if not isinstance(module, FullyShardedDataParallel):
+            # the root must be wrapped
+            raise TypeError(
```
Comment (awaelchli) [inline src/lightning/fabric/strategies/fsdp.py:335]: Could you update the grad clipping test to include this case?

File: src/lightning/fabric/strategies/fsdp.py (hunk 1 above)
Comment (carmocca) [reply src/lightning/fabric/strategies/fsdp.py:378]: Do you mean in PyTorch?

File: src/lightning/fabric/strategies/fsdp.py (hunk 2 above)
Comment (carmocca) [reply src/lightning/fabric/strategies/fsdp.py:391]: Wouldn't that be misleading? It doesn't match what the code is doing

File: src/lightning/fabric/strategies/fsdp.py [hunk 7]
```diff
@@ -684,7 +696,7 @@ def _optimizer_has_flat_params(optimizer: Optimizer) -> bool:
     return any(isinstance(param, FlatParameter) for group in optimizer.param_groups for param in group["params"])
 
 
-def _get_sharded_state_dict_context(module: "FullyShardedDataParallel") -> _GeneratorContextManager:
+def _get_sharded_state_dict_context(module: Module) -> Generator[None, None, None]:
```
[the remark this answers is not in the document]
Comment (carmocca) [reply src/lightning/fabric/strategies/fsdp.py:699]: _GeneratorContextManager is protected API and PyCharm underlines the import with a bright red. The correct return is Generator, but mypy has a bug and doesn't support it. I would favor correctness and simplicity over avoiding the ignore comment

File: src/lightning/fabric/strategies/fsdp.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning/fabric/strategies/fsdp.py:412]: Since you added this explicit error, I assume we are missing the proper handling? We could open an issue on our side to investigate.

File: tests/tests_fabric/strategies/test_fsdp.py [hunk 8]
```diff
@@ -302,13 +308,13 @@ def test_fsdp_load_checkpoint_one_fsdp_module_required(tmp_path):
     # missing FSDP model
     with pytest.raises(ValueError, match="Could not find a FSDP model in the provided checkpoint state."):
         strategy.load_checkpoint(path=tmp_path, state={"other": "data"})
-    with pytest.raises(ValueError, match="Could not find a FSDP model in the provided checkpoint state."):
-        strategy.load_checkpoint(path=tmp_path, state={"model": torch.nn.Linear(3, 3)})
```
Comment (awaelchli) [inline tests/tests_fabric/strategies/test_fsdp.py:306]: Why is this removed? I expect there to be an error, because there is no FSDP (sub)module in the state.

File: tests/tests_fabric/strategies/test_fsdp.py [hunk 9]
```diff
@@ -273,13 +279,13 @@ def test_fsdp_save_checkpoint_one_fsdp_module_required(tmp_path):
     # missing FSDP model
     with pytest.raises(ValueError, match="Could not find a FSDP model in the provided checkpoint state."):
         strategy.save_checkpoint(path=tmp_path, state={})
-    with pytest.raises(ValueError, match="Could not find a FSDP model in the provided checkpoint state."):
```
Comment (awaelchli) [inline tests/tests_fabric/strategies/test_fsdp.py:276]: Same here

File: src/lightning/fabric/strategies/fsdp.py (hunk 1 above)
Comment (carmocca) [reply src/lightning/fabric/strategies/fsdp.py:413]: Opened https://github.com/pytorch/pytorch/issues/105379

File: tests/tests_fabric/strategies/test_fsdp.py (hunk 8 above)
Comment (carmocca) [reply tests/tests_fabric/strategies/test_fsdp.py:306]: I simply forgot to add this back

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
Glad it was relatively uncomplicated to do this! Great work

File: src/lightning/fabric/strategies/fsdp.py [hunk 10]
```diff
@@ -255,30 +251,32 @@ def setup_module_and_optimizers(
...
-            module=module,
-            cpu_offload=self.cpu_offload,
-            mixed_precision=self.mixed_precision_config,
-            sharding_strategy=self.sharding_strategy,
-            device_id=self.root_device.index,
-            **self._fsdp_kwargs,
-        )
+        from torch.distributed.fsdp import FullyShardedDataParallel
+
+        if any(isinstance(mod, FullyShardedDataParallel) for mod in module.modules()):
```
Comment (justusschock) [inline src/lightning/fabric/strategies/fsdp.py:259]: Does `module.modules()` also include the root module or only its children? If it includes both, feel free to resolve, otherwise we need to check the root module as well :)

**Review by justusschock — APPROVED** [review]
Looks good to me, only one question :)

File: src/lightning/fabric/strategies/fsdp.py (hunk 10 above)
Comment (awaelchli) [reply src/lightning/fabric/strategies/fsdp.py:259]: Yes, modules() returns all submodules recursively, including the root module itself.
